Added basic gio::Initable implementation

xanathar opened the pull request:
This adds a basic implementation of `gio::Initable` which can be used to allow objects to implement `GInitable`.

Limitations:
- It does not attempt to solve the current issue with graceful failure of object creation on the Rust side (https://github.com/gtk-rs/gtk-rs-core/issues/22)
- It does not wrap the helper functions (`g_initable_new`, `g_initable_new_valist` and `g_initable_newv`); they are left commented as generated by gir, in the same vein of other types already in gtk-rs
- Unit test is implemented only using the ffi functions for now

bilelmoussaoui commented on gio/src/auto/initable.rs:
```diff
@@ -0,0 +1,67 @@
+}
+
+impl Initable {
+    //#[doc(alias = "g_initable_new")]
+    //pub fn new<P: IsA<Cancellable>>(object_type: glib::types::Type, cancellable: Option<&P>, error: &mut glib::Error, first_property_name: Option<&str>, : /*Unknown conversion*//*Unimplemented*/Fundamental: VarArgs) -> glib::Object {
+    //    unsafe { TODO: call ffi:g_initable_new() }
+    //}
+
+    //#[doc(alias = "g_initable_new_valist")]
+    //pub fn new_valist<P: IsA<Cancellable>>(object_type: glib::types::Type, first_property_name: &str, var_args: /*Unknown conversion*//*Unimplemented*/Unsupported, cancellable: Option<&P>) -> Result<glib::Object, glib::Error> {
+    //    unsafe { TODO: call ffi:g_initable_new_valist() }
+    //}
```
You can disable both new_valist/new as they are not bindings friendly. Probably rename newv to new as well

xanathar replied:
Done in new version

sdroege commented on gio/src/auto/initable.rs:
```diff
@@ -0,0 +1,67 @@
+}
+
+impl<O: IsA<Initable>> InitableExt for O {
+    fn init<P: IsA<Cancellable>>(&self, cancellable: Option<&P>) -> Result<(), glib::Error> {
```
Calling this multiple times is not necessarily safe AFAIU from the docs

sdroege replied:
So maybe this needs to be `unsafe` (and that's why having a safe `new()` would be useful). Not entirely sure.

xanathar replied:
Yes, it's not safe to be called more than once (or anything other than exactly once).
It's not clear to me however how to enforce that: in pure Rust I would have used a type-state / builder-object pattern kind of thing where the init takes ownership of self and returns a type which does not support Initable anymore, but in this interop scenario with the gobject model I'm not sure of the approach to take.

sdroege replied:
You mark it `unsafe` and add a `new()` that internally calls it :)

xanathar replied:
I wrote my message in between your first two and didn't refresh :) it was actually pretty clear what to do

xanathar replied:
Done. I called it `initable_new` to avoid name conflicts with `new` which is in widespread use, but I can easily change that.

sdroege commented on gio/src/auto/initable.rs:
```diff
@@ -0,0 +1,67 @@
+
+impl Initable {
+    //#[doc(alias = "g_initable_new")]
+    //pub fn new<P: IsA<Cancellable>>(object_type: glib::types::Type, cancellable: Option<&P>, error: &mut glib::Error, first_property_name: Option<&str>, : /*Unknown conversion*//*Unimplemented*/Fundamental: VarArgs) -> glib::Object {
```
You could implement this manually around `glib::Object::new()` and `Initable::init()` if you want. That should be like 5 lines of code :)

sdroege commented on gio/src/subclass/initable.rs:
```diff
@@ -0,0 +1,177 @@
+use crate::Cancellable;
+use crate::Initable;
+
+pub trait InitableImpl: ObjectImpl + Send {
```
```suggestion
pub trait InitableImpl: ObjectImpl {
```

`Initable` does not require the object to be send'able (and GObjects are never only `Send`, if anything they must be `Send + Sync` due to refcounting).

sdroege commented on gio/src/subclass/initable.rs:
```diff
@@ -0,0 +1,177 @@
+            let result = crate::ffi::g_initable_init(
+                test as *mut crate::ffi::GInitable,
+                std::ptr::null_mut(),
+                &mut error,
```
Maybe also assert that the error is null afterwards

sdroege commented on gio/src/subclass/initable.rs:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    #[test]
+    fn test_initable_through_ffi() {
```
You could also add another test that goes via the safe Rust API only

sdroege commented:
Looks good to me otherwise, thanks!

sdroege commented:
The CI failure will be gone if you rebase on latest master

sdroege commented on gio/src/initable.rs:
```diff
@@ -0,0 +1,34 @@
+
+pub trait InitableExtManual: Sized + 'static {
+    /// Helper function to safely create and initialize an object implementing [Initable].
+    fn initable_new<P: IsA<Cancellable>>(
```
You'd put that into an `impl Initiable` block, and then

```suggestion
    fn initable_new<T: IsA<Initable>, P: IsA<Cancellable>>(...) -> Result<T, Error>
```

sdroege replied:
Or maybe not... mh!

xanathar replied:
The reason I did it in an external trait (`InitableExtManual`) was to preserve the generation of the base interface from gir. I can move the whole Initable to manual if preferred, though.

sdroege replied:
You can keep the autogenerated code and have a separate `impl Initiable` in `src/initable.rs`. That's not a problem :)

sdroege commented on gio/src/subclass/initable.rs:
```diff
@@ -0,0 +1,215 @@
+    impl InitableTestType {
+        #[allow(clippy::new_without_default)]
+        pub fn new() -> Self {
+            glib::Object::new(&[]).expect("Failed creation of InitableTestType object")
```
This should go via the initable interface

xanathar replied:
I use the type in a couple of tests (`test_initable_with_init` and `test_initable_through_ffi`) which need the object in a non-initialized state. I can remove the tests, but I think the FFI one is good to check that a declaration of a Rust initable type would work called from C

xanathar replied:
`test_initable_with_initable_new` does the test using the initable interface

sdroege replied:
Ok then, but please add a comment about this :)

xanathar commented:
I pushed a new version which addresses the comments.
I accidentally pushed an intermediate one before rebasing, sorry, so please give the CI time to run twice 🤷🏻‍♂️

xanathar commented:
BTW - the CI failures are not bogus - there are a few types for which gir now adds "Initable" as an implemented interface. Tomorrow I'll re-run gir and push an update.

xanathar commented:
Exercising CI for minor findings.. as soon as it's green I will convert to "ready".
